fastchat/conversation.py

```diff
@@ -400,11 +470,135 @@ def to_openai_api_messages(self):
-            for role, message in self.messages
+    def to_vertex_api_messages(self):
+        from vertexai.preview.generative_models import Image
+        import base64
+        import requests
+
+        if self.system_message == "":
+            ret = []
+        else:
+            ret = [self.system_message]
+
+        for role, msg in self.messages[self.offset :]:
+            if msg is not None:
+                if type(msg) is tuple:
+                    text, images = msg[0], msg[1]
+                    for image in images:
+                        if image.startswith("http://") or image.startswith("https://"):
+                            response = requests.get(image)
+                            image = response.content
+                        else:  # base64
+                            image = base64.b64decode(image)
+                        ret.append(Image.from_bytes(image))
+                    ret.append(text)
+                else:
+                    ret.append(msg)
+
+        return ret
+
+    def to_anthropic_vision_api_messages(self):
+        """Convert the conversation to Claude-3 Messages Vision API format"""
+        ret = [
+            {
+                "role": "system",
+                "content": [{"type": "text", "text": self.system_message}],
+            }
         ]
+        for i, (_, msg) in enumerate(self.messages[self.offset :]):
+            if i % 2 == 0:
+                if type(msg) is tuple:
+                    content_list = [{"type": "text", "text": msg[0]}]
+
+                    for image_url in msg[1]:
+                        # Claude only supports base64
+                        if image_url.startswith("http://") or image_url.startswith(
+                            "https://"
+                        ):
+                            image_url = self.convert_image_to_base64(image_url)
+
+                        content_list.append(
+                            {
+                                "type": "image",
+                                "source": {
+                                    "type": "base64",
+                                    "media_type": "image/png",
+                                    "data": image_url,
+                                },
+                            }
+                        )
+
+                    ret.append({"role": "user", "content": content_list})
+                else:
+                    ret.append(
+                        {"role": "user", "content": [{"type": "text", "text": msg}]}
+                    )
+            else:
+                if msg is not None:
+                    ret.append(
+                        {
+                            "role": "assistant",
+                            "content": [{"type": "text", "text": msg}],
+                        }
+                    )
+        return ret
+
+    def save_new_images(self, use_remote_storage=False):
```

Comment:
@Ying1123 The big change after the rebase is how we save images since we have to take care of a new source of images: 1. URL 2. local source (supported as base64 str).

**Supporting URLs**: The remote storage PR allows us to store image files on GCS but we now have two issues: 1. persisting URLs in `conv.json` and 2. loading URLs into models later on. To solve (1), whenever I save an image, it will be converted from base64 to a GCS URL, then I save it into messages which means when we call `state.dict()` later on, we will extract this URL appropriately. For (2), some models only support base64 such as Gemini and Claude while GPT-4 supports URLs, in that case, I just convert the URL into base64 before loading into the model.

**Additional note:** In terms of saving images, I moved the logic to `conversation` because I only want to save the image once. Before, I believed we would always load in the image and convert it to a hash because the function to `conv.get_images()` would be true whenever there is one image. This would incur additional overhead for multi-turn multimodal conversations.